skiko/src/commonMain/kotlin/org/jetbrains/skiko/OverCommitmentBarrier.kt

```diff
@@ -0,0 +1,30 @@
+package org.jetbrains.skiko
+
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.channels.Channel
+
+/**
+ * Barrier for preventing more than [MAX_INFLIGHT_COMMAND_BUFFERS] being commited at the same time.
+ * It prevents situation where CPU encodes more commands than GPU can process (GPU bottleneck)
+ */
+class OverCommitmentBarrier {
+    private val channel = Channel<Unit>(capacity = MAX_INFLIGHT_COMMAND_BUFFERS, onBufferOverflow = BufferOverflow.DROP_OLDEST)
+
+    init {
+        repeat(MAX_INFLIGHT_COMMAND_BUFFERS) {
+            signal()
+        }
+    }
+
+    fun signal() {
+        channel.trySend(Unit)
+    }
+
+    suspend fun await() {
+        channel.receive()
+    }
+
+    companion object {
+        const val MAX_INFLIGHT_COMMAND_BUFFERS = 3
```

Comment:
MTLCommandBuffer is a transient object in Metal, they are not reused after being committed. It's max amount indeed related to triple buffering.
value of 1 would is basically making us wait until MTLCommandBuffer with presentation command is finished before starting a new one, making GPU and CPU work in sync, which we don't want.
value of 2 makes it double buffering, which doesn't utilise our CAMetalDrawable swap chain capabilities.
Hence 3.

Reply:
Does I understand it correctly?
```
layer.nextDrawable - returns drawable
layer.nextDrawable - returns drawable
layer.nextDrawable - returns drawable
layer.nextDrawable - returns null

commandBuffer.commit - sends drawable to GPU, after completion it returns to the layer pool
commandBuffer.commit - add drawable to queue, but don't send it to GPU yet
commandBuffer.commit - remove drawable from queue, return it to the pool, add the other drawable to queue
``

Reply:
It blocks if drawable pool is exhausted. Exhaustion of drawable pool is implementation detail of iOS and macOS, it guarantees to have at least 3 drawables to use in case we [don't change it to 2](https://developer.apple.com/documentation/quartzcore/cametallayer/2938720-maximumdrawablecount?language=objc). As you could see from examples, relying on it as a mechanism to prevent overcommitment brings some chaos.

Reply:
> Do I understand it correctly?
> 
> ```
> layer.nextDrawable - returns drawable
> layer.nextDrawable - returns drawable
> layer.nextDrawable - returns drawable
> layer.nextDrawable - returns null
> 
> commandBuffer.commit - sends drawable to GPU, after completion it returns to the layer pool
> commandBuffer.commit - add drawable to queue, but don't send it to GPU yet
> commandBuffer.commit - remove drawable from queue, return it to the pool, add the other drawable to queue
> ```

Not exactly. The drawable is captured at the moment `nextDrawable` is called. CommandBuffer (by default) retains all the resources which are used throughout the frame (textures, buffers, drawable) once commited. When you commit a commandBuffer, it simply goes into the queue of GPU execution. GPU executes commands without any signalling and waiting, we can consider GPU a parallel thread. CommandBuffer releases(in ARC terms) its resources once it's finished (completion handler is called).

Reply:
Generally we want to acquire drawable [**as late as possible**](https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/MTLBestPracticesGuide/Drawables.html#//apple_ref/doc/uid/TP40016642-CH2-SW1) in the frame. That means, that all work which doesn't require real drawing surface (off-screen work), should be in a separate commandBuffer. Right now we have one more command buffer, which doesn't make sense, because the only thing second command buffers does - is presenting the drawable.

Since the drawable is acquired early in the frame and used inside Skia to issue command, that's redundant.

Reply:
> Drawable is acquired by nextDrawable and is released

Is releasing means returning to the pool?

What happens when we commit to fast? I.e. when we commit first time - it goes to GPU. But what happens when we commit immediately the second/third time?

Reply:
### https://developer.apple.com/documentation/quartzcore/cametallayer/2887087-displaysyncenabled?language=objc### is macOS only property, which is used to synchronise read-write during vsyncs from-to IOSurface:

drawable bytes -> IOSurface bytes-> physical screen bytes

It's like that AFAIK, because on MacOS all windows rendering is managed by a separate window manager, which uses IOSurface as interprocess-interaction medium.